File: Longest K unique characters substring - GFG/longest-k-unique-characters-substring.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GfG {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            String s = sc.next();
            int k = sc.nextInt();
            Solution obj = new Solution();
            System.out.println(obj.longestkSubstr(s, k));
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    public int longestkSubstr(String s, int k) {
        Map<Character, Integer> freq = new HashMap<>();
        int i = 0, j = 0;
        int n = s.length();
        int longest = -1;
        
        while(j < n){
            char cur = s.charAt(j);
            freq.put(cur, freq.getOrDefault(cur, 0) + 1);
            
            if(freq.size() > k){
                while(!freq.isEmpty() && i < j && freq.size() > k){
                    char c = s.charAt(i);
                    i++;
                    int count = freq.get(c);
                    if(count == 1){
                        freq.remove(c);
                    } else {
                        freq.put(c, count - 1);
                    }
                }
            }
            
            if(freq.size() == k){
                longest = Math.max(longest, j - i + 1); //current window size
            }
            
            j++;
        }
        
        return longest;
    }
}

File: Longest K unique characters substring - GFG/longest-k-unique-characters-substring.py
class Solution:
    def longestKSubstr(self, s, k):
        #To track the no of keys in dict
        count =0
        # It will store the longest substring.
        res = -1
        # To keep track of no of characters
        ds = {}
        #start of window
        i =0
        # length of string
        n = len(s)
        # if no of unique characters are greater than the lenght of string
        if k>n:
            return -1
            
        for j in range(n):
            if s[j] not in ds:
                while (count==k):
                    ds[s[i]] -=1
                        # if the count of any element becomes 0. decrement the key count
                    if (ds[s[i]] == 0):
                        del ds[s[i]]
                        count-=1
                    
                    i+=1
                # add the new element in dic with freq 1
                ds[s[j]] = 1
                # increment the key count
                count+=1
            else:
                ds[s[j]]+=1
            #calculate the longest substring.
            if count == k:
                res= max(res, j - i + 1)    
            
        #return -1 if no of unique characters are less than k
        return res 





#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':

    t = int(input())

    for _ in range(t):
        s = input()
        k = int(input())

        solObj = Solution()

        ans = solObj.longestKSubstr(s, k)

        print(ans)

# } Driver Code Ends